core/domain/event_services_test.py

```diff
@@ -16,20 +16,164 @@
 
 """Tests for event handling."""
 
+import importlib
+import inspect
+import os
+import re
+
 from core.domain import event_services
 from core.platform import models
 from core.tests import test_utils
 import feconf
 
 from google.appengine.ext import ndb
 
+(stats_models, feedback_models) = models.Registry.import_models([
+    models.NAMES.statistics, models.NAMES.feedback])
+
 taskqueue_services = models.Registry.import_taskqueue_services()
 
 
 class MockNumbersModel(ndb.Model):
     number = ndb.IntegerProperty()
 
 
+class BaseEventHandlerTests(test_utils.GenericTestBase):
+
+    def test_handle_event_raises_not_implemented_error(self):
+        with self.assertRaisesRegexp(
+            NotImplementedError,
+            re.escape(
+                'Subclasses of BaseEventHandler should implement the '
+                '_handle_event() method, using explicit arguments '
+                '(no *args or **kwargs).')):
+            event_services.BaseEventHandler.record()
+
+
+class ExplorationActualStartEventHandlerTests(test_utils.GenericTestBase):
+
+    def test_record_exploration_actual_start_events(self):
+        all_models = (
+            stats_models.ExplorationActualStartEventLogEntryModel.get_all())
+
+        self.assertEqual(all_models.count(), 0)
+
+        event_services.ExplorationActualStartEventHandler.record(
+            'exp_id', 1, 'state_name', 'session_id')
+
+        all_models = (
+            stats_models.ExplorationActualStartEventLogEntryModel.get_all())
+
+        self.assertEqual(all_models.count(), 1)
+
+        for model in all_models:
+            self.assertEqual(model.exp_id, 'exp_id')
+            self.assertEqual(model.state_name, 'state_name')
+            self.assertEqual(model.session_id, 'session_id')
+            self.assertEqual(model.exp_version, 1)
+
+
+class SolutionHitEventHandlerTests(test_utils.GenericTestBase):
+
+    def test_record_solution_hit_events(self):
+        all_models = (
+            stats_models.SolutionHitEventLogEntryModel.get_all())
+
+        self.assertEqual(all_models.count(), 0)
+
+        event_services.SolutionHitEventHandler.record(
+            'exp_id', 1, 'state_name', 'session_id', 2.0)
+
+        all_models = (
+            stats_models.SolutionHitEventLogEntryModel.get_all())
+
+        self.assertEqual(all_models.count(), 1)
```

Comment:
Since the model count is 1, just de-indent next for loop and use a temporary `model` variable instead (here and elsewhere).

Reply:
This is required as `all_models` is of 'Query' type and 'Query' object does not support indexing

Reply:
In that case, use `model = next(all_models)`

Reply:
all_models is not an iterator so I did list(all_models)[0]. That worked! And I also read before that we should always do list(models) in our codebase and work with the list instead so as to not hit the database much often. Do you think we should implement this in our codebase?

Reply:
Where did you read that? I don't remember that being a rule, and in fact it seems like an antipattern if one is intending to retrieve just the first few models (you don't want to fetch the first 10000 of them and make a list if you only care about the top 5).

Reply:
I mean whatever models you are fetching just convert that Query into a list, that way it would be faster

I remember I heard this from Steve Huffman (the reddit guy), but whatever you say is fine actually. :)

Reply:
`Query` classes are indeed iterable, see the documentation: https://cloud.google.com/appengine/docs/standard/python/ndb/queries#iterators. Try `next(iter(all_models))` instead.

Reply:
I tried next(all_models) and there was an error. Is there any problem with list(all_models)[0]?

Reply:
Actually, looks like [`Query.get`](https://cloud.google.com/appengine/docs/standard/python/datastore/queryclass#Query_get) is exactly what we need here.

So try using `all_models.get()` here.

Reply:
Yes, `list(all_models)` will force every single result in a query to resolve. It's more efficient and a better pattern to take what we need by taking advantage of iterators (`next(iter(all_models))` or `all_models.get()` which is equivalent).

Reply:
So is it fine to use list or should I use get()?

Reply:
Use `get()`.

Reply:
Done!